### DISCUSSION TITLE
Soft AP mode enabling

### DISCUSSION THREAD

**Discussion by martusi61** [arendst/Tasmota#13362] [Ideas]
It may be useful to activate the function soft AP for direct connection with a wifi device such as smartphone or other. In this way, the possibilities of using the 8266 with Tasmota would be expanded, being able to directly control devices such as for example in the automotive field.

**Comment by sfromis** [comment]
Soft AP mode is already used for initial configuration, with no plans for making a "full" standalone mode. Tasmota should be able to connect to a smartphone running a wifi hotspot.

**Reply by martusi61** [reply]
Yes, I know, but my idea was precisely to free yourself from a hotspot, via a direct smartphone-esp8266 connection, in order to directly control a device

**Reply by sfromis** [reply]
Suggested quite a few times. It is at odds with how Tasmota is structured, and nothing has been done.

**Reply by martusi61** [reply]
Too bad, because it could open up new uses. Thanks and good job

**Reply by Fierelier** [reply]
@sfromis I hate using my phone and the ESP32-C3 does a horrible job at connecting to both my wifi hotspots, one being pure 2.4ghz, the other being separated 2.4ghz/5ghz SSIDs (no band steering). Meanwhile, the plug works great as an AP.

Why not simply add a "Use as AP instead" button to the wifi menu? You could always choose to connect it to a wifi device at a later date.

[My device](https://templates.blakadder.com/eightree_16A.html) is basically impossible to take apart, and I had almost bricked it using my vibecoded crap. I'm happy the janky code I have works, mostly -- I'm just gonna leave it as is for fear of breaking it.

It'd be nice to have a sanctioned way of doing this. -- Although honestly, it's my fault for getting such a physically closed off device (assuming there are any that are easy to open), otherwise I'd probably try to properly implement it myself.

**Comment by GregESP19** [comment]
Hello,
AP mode works fine with AIThinker firmware (AiThinker_ESP8266_DIO_8M_8M_20160615_V1.5.4.bin)

**Reply by martusi61** [reply]
thanks, i will try it, even if the possibilities of tasmota are unmatched

**Reply by GregESP19** [reply]
I use Wifi Controller ESP8266 (https://play.google.com/store/apps/details?id=com.mightyit.gops.wificontroller&hl=fr&gl=US) for control the relays (Switch on / off)

You can switch on / off the relay under linux with commands : 

echo -en "\xA0\x01\x01\xA2" | nc -w 0 192.168.4.1 8080
echo -en "\xA0\x01\x00\xA1" | nc -w 0 192.168.4.1 8080

echo -en "\xA0\x02\x01\xA3" | nc -w 0 192.168.4.1 8080
echo -en "\xA0\x02\x00\xA2" | nc -w 0 192.168.4.1 8080

**Reply by martusi61** [reply]
Thank you very much

**Reply by martusi61** [reply]
Thank you very much

**Reply by fb-pilot** [reply]
figuring out a solution for my mobile home I just read the conversation about the "soft-AP" and love this idea ! I use Tasmota in several applications and would prefer to use it in my mobile home too. But installing a router HW (which is not able to connect to anywhere)  only to have access to Tasmota now and then makes me not happy.
I understand that it would have some implication to Tasmota itself - but would be a nice feture for me. :-)

**Reply by pkkrusty** [reply]
Great for remote sensors too. Tasmota has expanded way beyond its original scope, with drivers for all sorts of sensors. The ability to turn off MQTT (setoption3) is technically not in scope either. I think it's a fair use case. Tasmota integrating scripting, sensors, and rules under one easy to use GUI makes it such a simple choice for beginners and experts to use. I agree that it's a worthy feature request. 

Android smartphone hotspots assign a different subnet on each connection (making it somewhat difficult to reach the device), and iOS does not allow easy connection to hotspot connected devices either (no IP address assignments listed).

**Reply by fb-pilot** [reply]
Hi all,
I just did a shor test with a Wemos D1 (ESP8266)
The feature "Wi-Fi Range Extender " (you have to compile it) provides the desired functionality of the soft-AP as far as I can see !

**Reply by philfleck** [reply]
Thanks for the info, i will try it report back. If that works, obviously we dont need my solution.

**Reply by pkkrusty** [reply]
Quick test with a ESP32 seems to work also. Great suggestion! Will have to do some additional testing once I pull the ESP32 out of wifi range to see how it reacts.

**Reply by fb-pilot** [reply]
in my tests I struggled with the reconnection, when the Client or the AP is rebooted ...
and I get an emty page instead of "Information" (.../in?)....
... further researching ...

**Reply by pkkrusty** [reply]
When the primary Wifi AP is not available, the Soft AP does not start. I can connect via serial and enter rgxstate 0 and then rgxstate 1 and the Soft AP will not start transmitting.  Might be a simple hack in the driver to allow this, but currently not the solution we're looking for.

**Reply by pkkrusty** [reply]
In the meantime, here's the altered webserver.ino that works with 12.4.0.4.  Don't need to alter support_tasmota.ino. 
[xdrv_01_9_webserver.ino.zip](https://github.com/arendst/Tasmota/files/11050545/xdrv_01_9_webserver.ino.zip)

**Reply by fb-pilot** [reply]
ok - super
did not test that situation ....
thank you for the info....
we should have this in the Tasmota-standard - e.g activated with an #ifdef !!

**Reply by pkkrusty** [reply]
I tried entering a wifi network with hack AP mode always enabled, and on startup, tasmota looks for it for about 10 seconds then activates the Soft AP and things work 'normally'. However, if I activate my phone's hotspot (the wifi network I entered) and it's visible to Tasmota on startup, it will connect. However, webserver does not work on the hotspot assigned IP or the Soft AP (which does not activate). However, Tasmota is still functional for NTP requests, MQTT, etc. I wonder if there's a simple change we can make to the webserver driver that will activate webserver correctly when Tasmota starts and finds the wifi hotspot. I haven't dug into the code yet, but @philfleck might have your head wrapped around it already.

**Reply by philfleck** [reply]
> In the meantime, here's the altered webserver.ino that works with 12.4.0.4. Don't need to alter support_tasmota.ino. [xdrv_01_9_webserver.ino.zip](https://github.com/arendst/Tasmota/files/11050545/xdrv_01_9_webserver.ino.zip)

tested on multiple devices,works!

**Reply by philfleck** [reply]
> ... I haven't dug into the code yet, but @philfleck might have your head wrapped around it already.


Theoretically it should work, but i discovered is, that once a valid wifi is entered you cant go back to AP mode. However, ones the wifi goes way, the device will keep searching for wifi. In our case, we keep it in standalone and only connect to it, since it is remote anyway.

**Reply by pkkrusty** [reply]
I like the idea of having the remote unit available for maintenance (installing a watering switch for someone, for example). I can ask them to turn on their hotspot (pre-programmed in the firmware), restart the unit, and it connects to their phone and activates MQTT. Then I can control it and check status remotely (via MQTT). This works until they turn off hotspot and restart the device. Then it's back to normal Soft AP mode. (It searches for the coded AP for 10 seconds before giving up and starting the Soft AP). 

The problem I'd like to solve is the webserver not activating correctly when the device joins the hotspot. Would solve the problem of restarting an inaccessible unit.

**Reply by pkkrusty** [reply]
Here's the 12.5.0.1 xdrv_01_9_webserver.ino, now altered such that all you have to do is `#define USE_ALWAYS_AP` in user_config_override.h to switch modes. 
[xdrv_01_9_webserver.ino.zip](https://github.com/arendst/Tasmota/files/11269296/xdrv_01_9_webserver.ino.zip)

**Reply by philfleck** [reply]
wow, thanks for the effort, that already feels like a feature @Jason2866 ;)
i sadly cannot allocate much time for this on my schedule. thanks again @pkkrusty.

**Reply by sfromis** [reply]
I'd say that all those ifs and buts make it **not** feel like a "feature", but instead what it is, a hack.

**Reply by philfleck** [reply]
this is likely true for a number of things that do not come with the prebuilts. what would it take to turn into a legit feature instead of a hack?

**Reply by pkkrusty** [reply]
Probably for someone with a deeper understanding of the wifi driver and webserver driver to integrate it a bit better. I'll keep messing with it a bit and see if I can solve the problem when joining an in-range wifi network. But primary developers don't seem interested in taking the time for it. Not a huge deal, since it's pretty easy to implement the hack for the few of us that find it useful.

**Reply by pkkrusty** [reply]
Digging into the code a bit, the Startwebserver function needs a mismatch between Web.state and type to activate the webserver correctly, so an #ifdef making Web.state HTTP_NONE at the beginning of the function enables the webserver correctly. However, because that function runs once a second, need to find a flag that lets it happen once but not more than that. Will keep playing with it.

Next problem comes when the hotspot is disconnected. Once again, the web manager function needs to re-activate the webserver with the 192.168.4.1 IP address. Not a huge deal, since the device can just be power cycled and if the hotspot isn't available it activates soft-ap after 20 seconds. But would like to figure it out without a restart.

**Reply by pkkrusty** [reply]
Made some progress.  Now, if hotspot is available on startup, Tasmota will join, and activate webserver correctly. When hotspot is turned off, Tasmota will start scanning for it, and after 20 seconds, re-activate the soft-AP plus webserver. 

So webserver is available in all modes. Again, need to
 ```
#define USE_ALWAYS_AP
#define WIFI_CONFIG_TOOL       WIFI_MANAGER
#define WIFI_SCAN_AT_RESTART   false
#define WIFI_SCAN_REGULARLY    false
```
to make things work smoothly.

I've only tested this on ESP8266. Note that in the weblog `WifiManager active for 3 minutes` doesn't mean anything any more. It will not restart, even if you have wifi credentials entered.

[xdrv_01_9_webserver.ino.zip](https://github.com/arendst/Tasmota/files/11291022/xdrv_01_9_webserver.ino.zip)

**Comment by GregESP19** [comment]
![IMG_20220824_095058_resize_29](https://user-images.githubusercontent.com/111926581/186405255-e40bf3b7-b432-4cee-b53a-abaafd5e1d00.jpg)

**Comment by philfleck** [comment]
Hi all,

i had a look throughout the forum, there are two other discussions related to tasmota standalone AP. 
As mentioned here, the main usecase is likely for remote location with no or no persistent wifi, connecting via smartphone.

i only tested with a esp32 modul and it remains responsive in AP mode.
My main fix was to force the HTTP_ADMIN and to call BUTTON_MAIN.

MQTT is turned off.

[xdrv_01_9_webserver.ino.txt](https://github.com/arendst/Tasmota/files/10705679/xdrv_01_9_webserver.ino.txt)
[my_user_config.h.txt](https://github.com/arendst/Tasmota/files/10705680/my_user_config.h.txt)

please link that to the other topics.

best g

**Reply by pkkrusty** [reply]
Which firmware version  are these modded from?

**Reply by barbudor** [reply]
There are a few forgotten problems here
Tasmota design requires to have a proper date and time so without NTP access either a external RTC clock or a GPS is required

**Reply by meingraham** [reply]
> There are a few forgotten problems here Tasmota design requires to have a proper date and time so without NTP access either a external RTC clock or a GPS is required

I believe you can execute a TIME command with a rule at boot to set a fake time to satisfy Tasmota.

**Reply by barbudor** [reply]
Hi @meingraham, long time no see 👋

Yes. But like the above it's more a hack than an intended feature
Also esp clock may not be very accurate (vs a RTC or NTP server) so do expect fast drift

**Reply by meingraham** [reply]
> it's more a hack than an intended feature

Indeed. But for an "edge" case (yes, less so than it used to be), it may suffice.

Back to my cave now 🤣

**Reply by philfleck** [reply]
just to make some points clear. i do not have a in depth knowledge of tasmota on which side-effects this might be introduced.

obviously all time related stuff like cron, date etc wont work

but, relative time does work

in my case, it is used on a custom esp32-board, to run a berry script watching some gpio. so far i can tell this works great

i also tested the esp32-cam, where i could control the led and it was responsive but i did not get any image.

i just commited it here, since i found a couple of users asking for that. i am pretty sure it is sufficient to control a switch, since in my case it is at a remote hard to reach location without wifi. Allowing for easy on/off function.

my thought was, since this easy fix worked, maybe there is a more tasmota way to do it . 
Maybe adding a userconfig define might help out some folks in need.

i used the master branch from Friday, and tried it on a hand full of boards.

hope that makes sense...

**Reply by pkkrusty** [reply]
Tested your code on ESP8266 and seems to work. Very helpful hack, thank you.
![IMG_23D86E7E7A70-1](https://user-images.githubusercontent.com/79770016/220314777-953cff36-5160-452a-9e79-7771f9c45eb0.jpeg)

@ascillato Could you maybe add a "work-around" tag to the discussion to visually indicate that a solution exists in the thread?

**Reply by philfleck** [reply]
@meingraham maybe a "fixed version" of this can be added to tasmota, so it is usable via a define for anyone who wants to build his one soft-ap enabled version?

**Reply by pkkrusty** [reply]
*Update* - If you're going to use the above linked altered xdrv_01_9_webserver.ino (12.3.1 version) with 12.4.0.2, you'll need to alter line 1580 in support_tasmota.ino to `if (!WifiIsInManagerMode()) { StartWebserver(Settings->webserver, WiFi.localIP()); }` instead of `if (!WifiIsInManagerMode()) { StartWebserver(Settings->webserver); }`

**Reply by philfleck** [reply]
obviously we cant be hot fixing for every new update, i still have the feeling it should be an option via a define if one wants to compile an soft-ap version ... i guess, this would be easy ... 

many thanks for the update on the changes, since we are running the builds remotely on a hard to access space, we will stay for a while our current compiled vesion

**Reply by sfromis** [reply]
Changing from a hack to a feature is more than just a define for having the hack in the project source code

**Reply by philfleck** [reply]
i know, more than you might think. but i have too little insights into the necessary parts of tasmota to do the necessary changes. 
best philipp

**Reply by Jason2866** [reply]
This discussion has the label `wont/cant fix`. So do not expect this hack will be added in source code.

**Comment by lalo-uy** [comment]
It happened to me, some tomes the telephone hotspots share 5he internet
acces, but do not allow diferet devices to communicate in the Wifi net.
Some kind of security.

**Comment by jeroenvermeulen** [comment]
I have created a patch based on https://github.com/arendst/Tasmota/discussions/13362#discussioncomment-4932672
It is created against Tasmota 13.1.0 but will probably work on other versions too.
[Tasmota-Soft-AP-mode.patch](https://github.com/arendst/Tasmota/files/12707062/Tasmota-Soft-AP-mode.patch)
Apply in Tasmota root dir with command:
```
git  apply  /path/to/Tasmota-Soft-AP-mode.patch
```

**Reply by Rqbln** [reply]
I tested your firmware with an Athom smart plug, and it appears to be partially functional. When I reset the plug, I can connect to it without encountering a captive portal. However, when attempting to connect to the IP address, I receive an `ERR_CONNECTION_REFUSED` error, rendering the plug unusable...
I'm also considering integrating MQTT and MDNS for my projects. Do you anticipate any issues with this?

**Reply by Jason2866** [reply]
MDNS will make trouble. It has serious bugs. It prevents running a device stable.
Tasmota has uptimes without MDNS counted in years. With MDNS you can be happy to reach days of uptime.

**Reply by Rqbln** [reply]
It will only be enabled for the wifi connection, that's all
I think my smart plug will survive :)

**Reply by Jan-II** [reply]
Hello, I'm new to tasmot and just learning. I successfully installed tasmot on sonoff and it works. But I would like to have access even without AP wifi. Can someone please explain to me how to install Tasmota-Soft-AP-mode.patch on my sonoff?
Thank you

**Reply by HellGL** [reply]
The patch does not work anymore with version 14.2.0. Is it possible to adapt it to the current version?
Thank you.

**Reply by pkkrusty** [reply]
The author of the patch can update with the following driver adjustment. Works fine for me on 14.2.0.1 (Note, the adjustment is the same since version 13, code didn't need to change)
[xdrv_01_9_webserver.ino.zip](https://github.com/user-attachments/files/16645718/xdrv_01_9_webserver.ino.zip)

Replace your xdrv_01_9_webserver.ino with this one, and in your user_config_override.h put 
```
#define USE_ALWAYS_AP
#define WIFI_CONFIG_TOOL       WIFI_MANAGER        // [WifiConfig] Default tool if Wi-Fi fails to connect (default option: 4 - 
#define WIFI_SCAN_AT_RESTART   false             // [SetOption56] Scan Wi-Fi network at restart for configured AP's
#define WIFI_SCAN_REGULARLY    false              // [SetOption57] Scan Wi-Fi network every 44 minutes for configured AP's

**Reply by Jan-II** [reply]
Thanks it works great

**Reply by FulvioSpelta** [reply]
My test:

- Downloaded 14.3.0 sources
- Overwritten [xdrv_01_9_webserver.ino](https://github.com/user-attachments/files/16645718/xdrv_01_9_webserver.ino.zip)
- Added the #define in "user_config_override.h"
- Compiled and flashed into a Sonoff Basic R2

It is running ok with AP always on and webadmin interface.
Thanks

**Reply by axel-GB** [reply]
HI on 14.4.1 do not work. xdrv_01_9_webserver.ino  from 14.3.0
Tasmota-Soft-AP-mode.patch  do not work too. 
Could be the password to AP stored in config  (changable from GUI)?

**Reply by axel-GB** [reply]
How can i use AP2 password like this
 // bool softAP(const char* ssid, const char* passphrase = NULL, int channel = 1, int ssid_hidden = 0, int max_connection = 4);
  WiFi.softAP(TasmotaGlobal.hostname, **HERE AP2 PASSWORD**, channel, 0, 1);

i cąnt get acces to saved password  ap2 
how ?

**Comment by belveder79** [comment]
Updated and tested on ESP32 couple of minutes (flashed the factory to get user space and WIFI credentials deleted on a non-virgin ESP), works for couple of minutes so far...

[tasmota_15.2.0.patch](https://github.com/user-attachments/files/24647243/tasmota_15.2.0.patch)

**Reply by belveder79** [reply]
Ahh yes, and here's another example where Tasmota in standalone mode can come in useful:

https://github.com/belveder79/NIUScooterHacking
